### PR TITLE
Bump Kubernetes dependencies to v0.31.x

### LINKED ISSUE #2948: Pods lose 1.31 fields during creation
So tonight I was working on playing around with https://github.com/kubernetes-sigs/dra-example-driver/blob/main/README.md#demo and Kueue.

All I am doing is following the demo to create a mock DRA driver.

I can run all the examples in this repo by running.

I then install kueue (0.8.0) (`kubectl apply --server-side -f https://github.com/kubernetes-sigs/kueue/releases/download/v0.8.0/manifests.yaml`) and then all the examples fail due to validation of resourceClaims.

Pre install of Kueue:

```
kehannon@kehannon-thinkpadp1gen4i:~/Work/dra-example-driver$ kubectl create -f demo/gpu-test1.yaml 
namespace/gpu-test1 created
resourceclaimtemplate.resource.k8s.io/single-gpu created
pod/pod0 created
pod/pod1 created
```

Install Kueue:

```
kubectl apply --server-side -f https://github.com/kubernetes-sigs/kueue/releases/download/v0.8.0/manifests.yaml
```

And then I run the same example (deleted before I install Kueue).
```
kehannon@kehannon-thinkpadp1gen4i:~/Work/dra-example-driver/demo$ kubectl create -f gpu-test1.yaml 
namespace/gpu-test1 created
resourceclaimtemplate.resource.k8s.io/single-gpu created
Error from server (Invalid): error when creating "gpu-test1.yaml": Pod "pod0" is invalid: spec.resourceClaims[0]: Invalid value: core.PodResourceClaim{Name:"gpu", ResourceClaimName:(*string)(nil), ResourceClaimTemplateName:(*string)(nil)}: must specify one of: `resourceClaimName`, `resourceClaimTemplateName`
Error from server (Invalid): error when creating "gpu-test1.yaml": Pod "pod1" is invalid: spec.resourceClaims[0]: Invalid value: core.PodResourceClaim{Name:"gpu", ResourceClaimName:(*string)(nil), ResourceClaimTemplateName:(*string)(nil)}: must specify one of: `resourceClaimName`, `resourceClaimTemplateName`
```

Not really sure what is going wrong here.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Bump Kubernetes dependencies to v0.31.x.

#### Which issue(s) this PR fixes:

Fixes #2948

#### Does this PR introduce a user-facing change?

```release-note
Support for Kubernetes 1.31
```

### CODE REVIEW

**Comment by tenzen-y** [comment]
/hold
We keep parking until the first patch release so that we can avoid the early bugs on Kubernetes and we can release a patch version easily.

File: hack/update-codegen.sh [hunk 1]
```diff
@@ -21,40 +21,34 @@ set -o pipefail
 GO_CMD=${1:-go}
 CURRENT_DIR=$(dirname "${BASH_SOURCE[0]}")
 KUEUE_ROOT=$(realpath "${CURRENT_DIR}/..")
-CODEGEN_PKG=$($GO_CMD list -m -mod=readonly -f "{{.Dir}}" k8s.io/code-generator)
-cd "${CURRENT_DIR}/.."
+KUEUE_PKG="sigs.k8s.io/kueue"
+CODEGEN_PKG=$(cd "${TOOLS_DIR}"; go mod download; $GO_CMD list -m -mod=readonly -f "{{.Dir}}" k8s.io/code-generator)
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply hack/update-codegen.sh:25]: maybe use a dependency in the Makefile to download the tools

File: cmd/kueuectl/app/create/create_resourceflavor_test.go [hunk 2]
```diff
@@ -249,6 +249,7 @@ func TestResourceFlavorCmd(t *testing.T) {
 		"shouldn't create resource flavor with dry-run client": {
 			rfName:  "rf",
 			args:    []string{"--dry-run", "client"},
+			wantRf:  &v1beta1.ResourceFlavor{},
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/kueuectl/app/create/create_resourceflavor_test.go:252]: Generated client was updated and now on error it's returning empty object.

File: hack/update-codegen.sh [hunk 3]
```diff
@@ -21,40 +21,34 @@ set -o pipefail
...
+  --output-dir "${KUEUE_ROOT}/client-go" \
+  --output-pkg "${KUEUE_PKG}/client-go" \
   --with-watch \
-  --with-applyconfig
+  --with-applyconfig \
+  "${KUEUE_ROOT}/apis"
 
 # We need to clean up the go.mod file since code-generator adds temporary library to the go.mod file.
-"${GO_CMD}" mod tidy
+"${GO_CMD}" mod tidy
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply hack/update-codegen.sh:54]: Maybe it's not correct comment for this case, but sometimes we need to update go.mod and go.sum after client generation. Most often this happens after updating the code-generator library.

File: test/integration/framework/framework.go [hunk 4]
```diff
@@ -134,6 +136,9 @@ func (f *Framework) StartManager(ctx context.Context, cfg *rest.Config, managerS
 				Port:    webhookInstallOptions.LocalServingPort,
 				CertDir: webhookInstallOptions.LocalServingCertDir,
 			}),
+		Controller: crconfig.Controller{
+			SkipNameValidation: ptr.To(true),
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply test/integration/framework/framework.go:140]: Ginkgo running suits concurrently and it's happening that Framework initialise manager and controllers a few times. So in this case we have an duplicate names error. I think for integration tests we can skip it.  And we also can be sure that we don't have duplicates on e2e tests.

File: hack/update-codegen.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply hack/update-codegen.sh:54]: But updating the code-generator library should be separate from running `update-codegen`

File: test/integration/framework/framework.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply test/integration/framework/framework.go:140]: Uhm... that sounds like a sign of a different problem. When running concurrently, we should probably have different managers. But we can investigate in another PR.

[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply test/integration/framework/framework.go:140]: > we should probably have different managers

Yes, it is. But [here](https://github.com/kubernetes-sigs/controller-runtime/blob/b901db121e1f53c47ec9f9683fad90a546688c3e/pkg/controller/name.go#L26-L27) using global variable for all managers.

File: hack/update-codegen.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply hack/update-codegen.sh:54]: Yes you're right. But generated code can import some additional libraries then we need to update go.mod and go.sum. For example:

https://github.com/kubernetes-sigs/kueue/blob/869ab97424fe5355b8f1b76150210d9e040c1126/client-go/applyconfiguration/utils.go#L21

File: hack/update-codegen.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply hack/update-codegen.sh]: You're right. It's not necessary. Thanks!

File: pkg/controller/core/clusterqueue_controller.go [hunk 5]
```diff
@@ -739,19 +739,18 @@ func (r *ClusterQueueReconciler) takeSnapshot(ctx context.Context) {
...
-	key, quit := r.snapshotsQueue.Get()
+	cqName, quit := r.snapshotsQueue.Get()
 	if quit {
 		return false
 	}
 
 	startTime := time.Now()
 	defer func() {
-		log.V(5).Info("Finished snapshot job", "key", key, "elapsed", time.Since(startTime))
+		log.V(5).Info("Finished snapshot job", "key", cqName, "elapsed", time.Since(startTime))
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply pkg/controller/core/clusterqueue_controller.go]: Done

File: hack/update-codegen.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply hack/update-codegen.sh]: I think you right, we should update it only one time on updating library. Removed it. Thanks.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Comment by kannon92** [comment]
Looking forward to this change. COuld you please update the PR to fix conflicts?

**Comment by mbobrovskyi** [comment]
> Looking forward to this change. Could you please update the PR to fix conflicts?

This PR on hold and we are waiting for the first release patch of k8s.

**Comment by kannon92** [comment]
Can you add 1.31 to the E2E_K8S_VERSION in Makefile-test.mk?

**Comment by mbobrovskyi** [comment]
> Can you add 1.31 to the E2E_K8S_VERSION in Makefile-test.mk?

Good point, thanks!

**Comment by kannon92** [comment]
1.31.1 was released. Can you rebase this and get this in?

**Comment by mimowo** [comment]
Is this just a cleanup though? I think it deserves a release note, as it has user-facing consequences - it fixes https://github.com/kubernetes-sigs/kueue/issues/2948, and allows support for JobSet 0.6.0. 

Tentatively updating, but feel free to change / discuss if I'm missing something.

/kind feature
/release-note-edit
```release-note
Support for Kubernetes 1.31
```

**Comment by mbobrovskyi** [comment]
Yeah, it started as a small cleanup for the code-generator and ended up as a major feature for Kubernetes 1.31 :)

**Comment by mbobrovskyi** [comment]
/hold cancel

Due to first patch for k8s dependencies released.

**Comment by mimowo** [comment]
Please also add  "Fixes #2948" in the description. I know this issue is closed already, but I think it would make the relationship easier to discover for future readers who find the PR in the release-notes.

**Pull request merged**